src/hera/workflow_template.py

```diff
@@ -0,0 +1,176 @@
+"""The implementation of a Hera workflow for Argo-based workflows"""
+import warnings
+from typing import Dict, Optional
+
+from argo_workflows.models import (
+    IoArgoprojWorkflowV1alpha1DAGTemplate,
+    IoArgoprojWorkflowV1alpha1Template,
+    IoArgoprojWorkflowV1alpha1WorkflowTemplate,
+    IoArgoprojWorkflowV1alpha1WorkflowTemplateSpec,
+    ObjectMeta,
+)
+
+from hera.security_context import WorkflowSecurityContext
+from hera.task import Task
+from hera.volumes import Volume
+from hera.workflow_template_service import WorkflowTemplateService
+
+# from uuid import uuid4
+
+
+# by default, `DeprecationWarning`s are silenced, this removes the warning from the filter so it
+# can be issued to users
+warnings.simplefilter('always', DeprecationWarning)
+
+
+class WorkflowTemplate:
+    """A workflowTemplate representation.
+    The WorkflowTemplate is used as a functional representation for a collection of tasks and
+    steps. The WorkflowTemplate is basically the same as a Workflow but with a template you don't
+    have to write the same steps, you can reuse it over and over.
+    Parameters
+    ----------
+    name: str
+        The workflowTemplate name. Note that the workflowTemplate initiation will replace underscores with dashes.
+    service: WorkflowService
+        A workflowTemplate service to use for submissions.
+        See `hera.v1.workflow_template_service.WorkflowTemplateService`.
+    parallelism: int = 50
+        The number of parallel tasks to run in case a task group is executed for multiple tasks.
+    service_account_name: Optional[str] = None
+        The name of the service account to use in all workflow tasks.
+    security_context:  Optional[WorkflowSecurityContext] = None
+        Define security settings for all containers in the workflow.
+    labels: Optional[Dict[str, str]] = None
+        A Dict of labels to attach to the Workflow object metadata
+    namespace: Optional[str] = 'default'
+        The namespace to use for creating the WorkflowTemplate.  Defaults to "default"
+    """
+
+    def __init__(
+        self,
+        name: str,
+        service: WorkflowTemplateService,
+        parallelism: int = 50,
+        service_account_name: Optional[str] = None,
+        labels: Optional[Dict[str, str]] = None,
+        namespace: Optional[str] = None,
+        security_context: Optional[WorkflowSecurityContext] = None,
+    ):
+        self.name = name  # f'{name.replace("_", "-")}-{str(uuid4()).split("-")[0]}'  # RFC1123
```

Comment:
I like this! It's a nice transition to using the user submitted name exclusively. I think Hera should still perform the replacement of underscores with dashes because of RFC1123, but not add the UUID. Thoughts?

Reply:
Absolutely right. Thanks for your suggestion.